File: 0070_climb_stairs/climb_stairs.cpp
#include <iostream>
#include <vector>

/**
 * Time: O(n)
 * Space: O(n) 
 */
int climb_stairs(int n)
{
    if (n == 1)
        return 1;
    vector<int> res(n, 0);
    res[0] = 1;
    res[1] = 2;
    for (int i = 2; i < n; i++)
        res[i] = res[i - 2] + res[i - 1];
    return res[n - 1];
}

File: 0070_climb_stairs/climb_stairs.rb
def climb_stairs(n)
    return 1 if n == 1
    dp = Array.new(n, 0)
    dp[0] = 1
    dp[1] = 2
    2.upto(n - 1) { |i| dp[i] = dp[i - 1] + dp[i - 2] }
    dp.last
end